scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1131,6 +1133,47 @@ class MappablePipeJoinEnrichment[T](pipe: TypedPipe[T]) {
 }
 
 /**
+ * This class is to enrich the forceToDisk operation on TypedPipes and allow
+ * us to potentially optimize these operations when they're chained together
+ * redundantly.
+ */
+case class ForceToDiskTypedPipe[T](typedPipe: TypedPipe[T]) extends TypedPipe[T] {
+  import Dsl._
+
+  override def forceToDisk: TypedPipe[T] = this
+
+  override def asPipe[U >: T](fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]): Pipe = {
+    val pipe = typedPipe.toPipe[U](fieldNames)(flowDef, mode, setter)
+    pipe.forceToDisk
+  }
+
+  override def cross[U](tiny: TypedPipe[U]): TypedPipe[(T, U)] =
+    forceBefore(typedPipe).cross(tiny)
+
+  override def flatMap[U](f: T => TraversableOnce[U]): TypedPipe[U] =
+    forceBefore(typedPipe).flatMap(f)
```

Comment:
So if I understand correctly, the `ForceToDiskTypedPipe` seems to be giving us:
1) Protection against a user calling multiple forceToDisk operations successively by mistake
2) Type encapsulation which we could use maybe in the future if we have a fancier optimizer to prune away some of these ForceToDisk calls that are redundant.
3) Current scenario: If a force2Disk is called followed by a cross / flatMap we check the prior pipe / operations to potentially optimize things. I think in this case if we do a physicalForceToDisk.flatMap or physicalForceToDisk.cross always, that shouldn't make things worse than they currently are. 

We could drop the `ForceToDiskTypedPipe` and just do the checking in `HashJoinable.getMappedPipe`. That's similar to @isnotinvain 's suggestion earlier of walking the DAG in HashJoinable. 

Let me know if this sounds good, I'll update the PR accordingly.